// ngx-date-picker/datepicker/datepicker.config.js
import { Injectable } from '@angular/core';
export var DatepickerConfig = (function () {
    function DatepickerConfig() {
        this.datepickerMode = 'day';
        this.startingDay = 0;
        this.yearRange = 20;
        this.minMode = 'day';
        this.maxMode = 'year';
        this.showWeeks = false;
        this.formatDay = 'DD';
        this.formatMonth = 'MMMM';
        this.formatYear = 'YYYY';
        this.formatDayHeader = 'dd';
        this.formatDayTitle = 'MMMM YYYY';
        this.formatMonthTitle = 'YYYY';
        this.onlyCurrentMonth = false;
        this.monthColLimit = 3;
        this.yearColLimit = 5;
        this.shortcutPropagation = false;
        this.ouputFormat = 'YYYY-MM-DDTHH:mm:ssZ';
        this.timeZones = [
            {
                label: '-12:00',
                zone: -12
            },
            {
                label: '-11:00',
                zone: -11
            },
            {
                label: '-10:00',
                zone: -10
            },
            {
                label: '-09:30',
                zone: -559
            },
            {
                label: '-09:00',
                zone: -9
            },
            {
                label: '-08:00',
                zone: -8
            },
            {
                label: '-07:00',
                zone: -7
            },
            {
                label: '-06:00',
                zone: -6
            },
            {
                label: '-05:00',
                zone: -5
            },
            {
                label: '-04:00',
                zone: -4
            },
            {
                label: '-03:30',
                zone: -198
            },
            {
                label: '-03:00',
                zone: -3
            },
            {
                label: '-02:00',
                zone: -2
            },
            {
                label: '-01:00',
                zone: -1
            },
            {
                label: '00:00',
                zone: 0
            },
            {
                label: '01:00',
                zone: 1
            },
            {
                label: '02:00',
                zone: 2
            },
            {
                label: '03:00',
                zone: 3
            },
            {
                label: '03:30',
                zone: 198
            },
            {
                label: '04:00',
                zone: 4
            },
            {
                label: '04:30',
                zone: 258
            },
            {
                label: '05:00',
                zone: 5
            },
            {
                label: '05:30',
                zone: 318
            },
            {
                label: '05:45',
                zone: 327
            },
            {
                label: '06:00',
                zone: 6
            },
            {
                label: '06:30',
                zone: 378
            },
            {
                label: '07:00',
                zone: 7
            },
            {
                label: '08:00',
                zone: 8
            },
            {
                label: '09:00',
                zone: 9
            },
            {
                label: '09:30',
                zone: 558
            },
            {
                label: '10:00',
                zone: 10
            },
            {
                label: '10:30',
                zone: 618
            },
            {
                label: '11:00',
                zone: 11
            },
            {
                label: '12:00',
                zone: 12
            },
            {
                label: '12:45',
                zone: 747
            },
            {
                label: '13:00',
                zone: 13
            },
            {
                label: '14:00',
                zone: 14
            }
        ];
    }
    DatepickerConfig.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    DatepickerConfig.ctorParameters = function () { return []; };
    return DatepickerConfig;
}());
