// ngx-date-picker/datepicker/datepicker.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DatePickerInnerComponent } from './datepicker-inner.component';
import { DatePickerComponent } from './datepicker.component';
import { DayPickerComponent } from './daypicker.component';
import { MonthPickerComponent } from './monthpicker.component';
import { YearPickerComponent } from './yearpicker.component';
import { DatepickerConfig } from './datepicker.config';
import { DropdownState } from './dropdown.state';
import { DropdownContainerComponent } from './dropdown-container.component';
import { ComponentLoaderFactory } from '../component-loader/component-loader.factory';
import { PositioningService } from '../positioning/positioning.service';
import { TimePickerComponent } from './timepicker.component';
import { TimezonePickerComponent } from './timezonepicker.component';
export var DatepickerModule = (function () {
    function DatepickerModule() {
    }
    DatepickerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule
                    ],
                    declarations: [
                        DatePickerComponent,
                        DatePickerInnerComponent,
                        DayPickerComponent,
                        MonthPickerComponent,
                        YearPickerComponent,
                        DropdownContainerComponent,
                        TimePickerComponent,
                        TimezonePickerComponent
                    ],
                    exports: [
                        DatePickerComponent,
                        DayPickerComponent,
                        TimePickerComponent,
                        TimezonePickerComponent
                    ],
                    providers: [
                        ComponentLoaderFactory,
                        PositioningService,
                        DatepickerConfig,
                        DropdownState
                    ],
                    entryComponents: [
                        DatePickerComponent,
                        DropdownContainerComponent
                    ]
                },] },
    ];
    /** @nocollapse */
    DatepickerModule.ctorParameters = function () { return []; };
    return DatepickerModule;
}());
