// ngx-date-picker/datepicker/dropdown-container.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { DropdownState } from './dropdown.state';
export var DropdownContainerComponent = (function () {
    function DropdownContainerComponent(_state) {
        var _this = this;
        this._state = _state;
        this.isOpen = false;
        this._subscription = _state.isOpenChange.subscribe(function (value) {
            _this.isOpen = value;
        });
    }
    Object.defineProperty(DropdownContainerComponent.prototype, "direction", {
        get: function () {
            return this._state.direction;
        },
        enumerable: true,
        configurable: true
    });
    DropdownContainerComponent.prototype.ngOnDestroy = function () {
        this._subscription.unsubscribe();
    };
    DropdownContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dropdown-container',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        style: 'display:block;position: absolute;z-index:2000'
                    },
                    template: "\n    <div [class.dropup]=\"direction === 'up'\"\n         [class.dropdown]=\"direction === 'down'\"\n         [class.show]=\"isOpen\"\n         [class.open]=\"isOpen\"><ng-content></ng-content></div>\n  "
                },] },
    ];
    /** @nocollapse */
    DropdownContainerComponent.ctorParameters = function () { return [
        { type: DropdownState, },
    ]; };
    return DropdownContainerComponent;
}());
